### PR TITLE
Merge kernel caching logic for CUDA and OpenCL backends

### PR DESCRIPTION
**Pull request by 9prady9**
Partially addresses tasks from #2857

### CODE REVIEW

File: src/backend/common/kernel_util.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+#include <string>
+
+template<typename T>
+std::string toString(T value);
+
+struct TemplateArg {
+    std::string _tparam;
+
+    TemplateArg(std::string str) : _tparam(str) {}
```
Comment (umar456) [inline src/backend/common/kernel_util.hpp:20]:
```suggestion
    TemplateArg(std::string str) : _tparam(std::move(str)) {}
```

File: src/backend/common/kernel_util.hpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ ********************************************************/
+
+#include <traits.hpp>
+
+#include <string>
+
+template<typename T>
+std::string toString(T value);
+
+struct TemplateArg {
```
Comment (umar456) [inline src/backend/common/kernel_util.hpp:17]: maybe define this in its own header?

File: src/backend/common/kernel_util.cpp [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+#include <string>
+
+using std::string;
+
+template<typename T>
+string toString(T value) {
+    return std::to_string(value);
+}
+
+template string toString<int>(int);
```
Comment (umar456) [inline src/backend/common/kernel_util.cpp:25]: why not use the std::to_string

File: src/backend/common/kernel_interface.hpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+#include <cstddef>
+#include <utility>
+
+namespace common {
+
+/// Kernel Interface that should be implemented by each backend
+template<typename Module, typename Kernel, typename Enqueuer, typename DevPtr>
+class KernelInterface {
+   public:
+    using ModuleType   = Module;
```
Comment (umar456) [inline src/backend/common/kernel_interface.hpp:21]: These using statements don't add anything. IMO

File: src/backend/common/kernel_interface.hpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    using KernelType   = Kernel;
+    using EnqueuerType = Enqueuer;
+    using DevPtrType   = DevPtr;
+
+    Module mProgram;
```
Comment (umar456) [inline src/backend/common/kernel_interface.hpp:26]: These should be private

File: src/backend/common/kernel_util.hpp (hunk 2 above)
Comment (9prady9) [reply src/backend/common/kernel_util.hpp:17]: Moving them in next commit.

**Pull request merged**